Refactor GraphGeodesicDistance

ulupo opened the pull request:
**Types of changes**
<!--
What types of changes does your code introduce? Put an `x` in all the boxes that apply.
-->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

**Description**
There are a couple of major problems with the current implementation of `GraphGeodesicDistance`:
1. The return type of `transform` is always ndarray, even when different distance matrix shapes imply that the ndarray is 1D. But in this case, output fed to the homology transformers will fail due to the current implementation of `check_point_cloud`, see https://github.com/giotto-ai/giotto-tda/blob/188b6755b7f567a49d0a15cb63492de29a81ab45/gtda/utils/validation.py#L260
2. The handling of zero entries, infinity entries and non-stored entries in the sparse case was quite opaque and some non-generic shortcuts were taken in the code to address most, but not all, cases.

This PR fixes both problems and introduces additional changes. In particular:
- The return type of `transform` is only ndarray if it can be turned into a 3D ndarray, else it is list.
- Scipy's [`shortest_path`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.csgraph.shortest_path.html) replaces scikit-learn's [`graph_shortest_path`](https://scikit-learn.org/stable/modules/generated/sklearn.utils.graph_shortest_path.graph_shortest_path.html). This is because it supports a wider range of algorithms, it supports masked arrays, it is better maintained, and it has "better" behaviour in the sparse case (see below). I have created a gist to exhibit the difference in behaviour and the new behaviour of `GraphGeodesicDistance`: https://gist.github.com/ulupo/83cc82ce83379ebda8fdfe846d0c06a5.
- A particular convention is clearly established:
  - if the input arrays are dense, then absent edges must be indicated by ``numpy.inf``;
  - zero edges in int or float arrays do not denote absent edges, but edges of length 0;
  - on the other hand, ``False`` edges in Boolean arrays do denote absent edges;
  - in the sparse case, non-stored values are interpreted as absent edges.
- The additional parameters `directed`, `unweighted` and `method` are made available, with the obvious meanings.

**Checklist**
- [x] I have read the [guidelines for contributing](https://giotto-ai.github.io/gtda-docs/latest/contributing/#guidelines).
- [x] My code follows the code style of this project. I used `flake8` to check my Python changes.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed. I used `pytest` to check this on Python tests.

Commit:
Refactor GraphGeodesicDistance

- Change return type to list for compatibility with homology transformers
- Use scipy's shortest_path instead of sklearn's graph_shortest_path
- Make clearer rules on the role of zero entries, infinity entries, and non-stored values
- Add parameters directed, unweighted, and method
- Support masked arrays
- Rewrite docstring

Commit:
Fix See Also in KNeighborsGraph and TransitionGraph

wreise commented on gtda/graphs/geodesic_distance.py:
```diff
@@ -52,18 +77,26 @@ class GraphGeodesicDistance(BaseEstimator, TransformerMixin, PlotterMixin):
-        X_distance[X_distance == 0] = np.inf  # graph_shortest_path returns a
-        # float64 array, so inserting np.inf does not change the type.
-        np.fill_diagonal(X_distance, 0)
+        if not issparse(X) and not isinstance(X, MaskedArray):
```
What happens with boolean X? Is it passed directly to `shortest_path` and treated correctly?

ulupo replied:
That's correct! Check e.g. the gist https://gist.github.com/ulupo/83cc82ce83379ebda8fdfe846d0c06a5

ulupo replied:
@wreise The reason for converting to a masked array when the input is a dense non-boolean one is to avoid the situation depicted in input cell 5 in that gist. Using `scipy` directly, any zeros will be interpreted as absent edges, and not as edges for which the cost is zero (which is more useful in our applications).

wreise reviewed (commented):
Looks good. Maybe one thing that is missing are tests for the different supported input types  (binary, sparse, dense).

ulupo committed:
Prevent using FW algorithms when some edges are zero

- Work around https://github.com/scipy/scipy/issues/12424
- Introduce user warnings when algorithm cannot be chosen automatically or set to FW
- Fix test ground truth
- Improves array conversion in GraphGeodesicDistance.transform

ulupo commented on gtda/graphs/geodesic_distance.py:
```diff
@@ -52,18 +81,40 @@ class GraphGeodesicDistance(BaseEstimator, TransformerMixin, PlotterMixin):
+        method_ = self.method
+        if not issparse(X):
+            diag = np.eye(X.shape[0], dtype=bool)
+            if np.any(~np.logical_or(X, diag)):
```
Here, I am looking for non-diagonal zeros.

wreise commented on gtda/graphs/geodesic_distance.py:
```diff
@@ -88,15 +92,29 @@ def __init__(self, n_jobs=None, directed=False, unweighted=False,
                 # Convert to a masked array with mask given by positions in
                 # which infs or NaNs occur.
-                X = masked_invalid(X)
+                if X.dtype != bool:
```
Suggestion: aggregate the nested `if`s into one?

ulupo replied:
Oh dear, I tried and developed a headache :) Let me try again.

ulupo committed:
Add more GraphGeodesicDistance unit tests

ulupo committed:
Fix linting

ulupo committed:
Remove some unnecessary spaces
